### PR TITLE
Update to TypeScript 4.1

### PR DESCRIPTION
**Pull request by wgoehrig**
Took a crack at updating to the latest version of TypeScript (4.1), mainly to make sure we don't fall too far behind.

There haven't been _too_ many breaking changes since 3.7, but there are a couple that came up repeatedly that iModel.js developers should now be aware of.  Specifically:

- **[abstract Members Can’t Be Marked async](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-1.html#abstract-members-cant-be-marked-async)**
Just leave out the `async` keyword here from now on.

- **[resolve’s Parameters Are No Longer Optional in Promises](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-1.html#resolves-parameters-are-no-longer-optional-in-promises)**
This helps seal up a number of type-safety holes, but has the unfortunate side effect of breaking things like: 
    ```ts
    await new Promise((resolve) => { /* in a callback somewhere... */ resolve(); });
    ```
    The solution here is to just add an explicit type parameter to the Promise constructor:
    ```ts
    await new Promise<void>((resolve) => { /* in a callback somewhere... */ resolve(); });
    ```
- **[Properties Overriding Accessors (and vice versa) is an Error](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#properties-overriding-accessors-and-vice-versa-is-an-error)**
It is no longer legal for derived classes to wrap a base class property with a getter if the base property is not implemented as a getter.  This can be usually be avoided by just refactoring the base property to instead use a private property behind public getters and setters.

- **[Operands for delete must be optional](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#operands-for-delete-must-be-optional)**
We `delete` properties in a number of tests - an easy fix is to explicitly use a `Partial<...>` type for whatever you're mutating.

### CODE REVIEW

File: core/backend/src/CloudStorageBackend.ts [hunk 1]
```diff
@@ -139,7 +139,7 @@ export class AzureBlobStorage extends CloudStorageService {
         const blockSize = 100 * 1024 * 1024;
         const concurrency = 1;
         const result = await Azure.uploadStreamToBlockBlob(Azure.Aborter.none, source, blocks, blockSize, concurrency, blobOptions);
-        return resolve(result.eTag);
+        return resolve(result.eTag ?? "");
```
Comment (wgoehrig) [inline core/backend/src/CloudStorageBackend.ts:142]: @swbsi Is this safe? Would `result.eTag!` be better, or should we throw an error if it's undefined?

File: core/ecschema-metadata/src/Editor/Editor.ts [hunk 2]
```diff
@@ -910,7 +910,7 @@ export namespace Editors {
       if (koqProps.name === undefined) return { errorMessage: `No name was supplied within props.` };
       const newKoQ = (await schema.createKindOfQuantity(koqProps.name)) as MutableKindOfQuantity;
       if (newKoQ === undefined) {
-        return { errorMessage: `Failed to create class ${name} in schema ${schemaKey.toString(true)}.` };
+        return { errorMessage: `Failed to create class ${koqProps.name} in schema ${schemaKey.toString(true)}.` };
```
Comment (wgoehrig) [inline core/ecschema-metadata/src/Editor/Editor.ts:913]: @calebmshafer This probably should have been an error before, but it was actually using a global `name` var from lib.dom.d.ts before.  That name var has now been deprecated, so I had to guess at what this was supposed to be...

File: core/backend/src/IModelDb.ts [hunk 3]
```diff
@@ -1144,7 +1144,7 @@ export abstract class IModelDb extends IModel {
           if (ret.error !== undefined)
             reject(new Error(ret.error.message));
           else
-            resolve(ret.result);
+            resolve(ret.result!);
```
Comment (wgoehrig) [inline core/backend/src/IModelDb.ts:1147]: We could probably avoid the need for this by making `ErrorStatusOrResult` in NativeLibrary.d.ts a union type, but that would need to be a separate PR...

File: core/geometry/src/test/geometry4d/MomentData.test.ts [hunk 4]
```diff
@@ -202,7 +202,7 @@ describe("MomentData", () => {
         GeometryCoreTestIO.showMomentData(allGeometry, rawSums, true, x0, y0, 0);
         // GeometryCoreTestIO.showMomentData(allGeometry, principalMoments, true, x0, y0, 0);
         if (g instanceof CurveCollection) {
-          const strokes = g.cloneStroked(strokeOptions);
+          const strokes = (g as CurveCollection).cloneStroked(strokeOptions);
```
Comment (wgoehrig) [inline core/geometry/src/test/geometry4d/MomentData.test.ts:205]: @EarlinLutz tsc was basically complaining that `g` could never be a `CurveCollection` before I added this type assertion... Is this condition not even hittable?  If it is, we may need to remove the type assertion and investigate why tsc thinks this is impossible.

File: core/geometry/src/test/serialization/BSIJSON.test.ts [hunk 5]
```diff
@@ -66,7 +66,7 @@ function isDifferentTypeName(obj: object, noisy: boolean = false): boolean {
   if (obj.constructor) {
     if (obj.constructor !== previousConstructor) {
       if (noisy)
-        console.log("First ", name);
+        console.log("First ", obj.constructor.name);
```
Comment (wgoehrig) [inline core/geometry/src/test/serialization/BSIJSON.test.ts:69]: @EarlinLutz This probably should have been an error before, but it was actually using a global `name` var from lib.dom.d.ts before.  That name var has now been deprecated, so I had to guess at what this was supposed to be logging...

File: presentation/components/src/presentation-components/tree/controlled/WithFilteringSupport.tsx [hunk 6]
```diff
@@ -82,9 +82,9 @@ export function DEPRECATED_controlledTreeWithFilteringSupport<P extends Controll
     return (
       <div className="filteredTree">
         <TreeComponent
-          nodeLoader={nodeLoader}
           nodeHighlightingProps={nodeHighlightingProps}
           {...strippedProps as CombinedProps}
+          nodeLoader={nodeLoader}
```
Comment (wgoehrig) [inline presentation/components/src/presentation-components/tree/controlled/WithFilteringSupport.tsx:87]: @grigasp  Is it safe to change the ordering for these?  tsc was pointing out that `CombinedProps` still contains non-nullable `nodeLoader` and `visibleNodes` props...

File: test-apps/ui-test-app/src/frontend/appui/widgets/demodataproviders/demoTableDataProvider.ts [hunk 7]
```diff
@@ -195,7 +195,7 @@ export const onDropTargetDrop = (args: DropTargetArguments<DemoTableDragDropType
     const dragRow: RowItem = {
       key: id,
       extendedData: {
-        id, label, icon, type, description, parentId, dataProvider, ...rest,
+        label, icon, type, description, parentId, dataProvider, ...rest, id,
```
Comment (wgoehrig) [inline test-apps/ui-test-app/src/frontend/appui/widgets/demodataproviders/demoTableDataProvider.ts:198]: @bsteinbk is it safe to change the order here?  Technically `rest` contains an id property that would have overwritten `id` before...

File: presentation/components/src/presentation-components/tree/controlled/WithFilteringSupport.tsx (hunk 6 above)
Comment (wgoehrig) [reply presentation/components/src/presentation-components/tree/controlled/WithFilteringSupport.tsx:87]: see also: presentation/components/src/presentation-components/tree/controlled/WithVisibleNodes.tsx

File: ui/framework/src/ui-framework/shared/ItemDefBase.ts [hunk 8]
```diff
@@ -35,7 +35,9 @@ export abstract class ItemDefBase {
   /** @deprecated */
   public isEnabled: boolean = true;
   public isPressed: boolean = false;
-  public isActive: boolean = false;
+  private _isActive: boolean = false;
+  public get isActive(): boolean { return this._isActive; }
+  public set isActive(v: boolean) { this._isActive = v; }
```
Comment (wgoehrig) [inline ui/framework/src/ui-framework/shared/ItemDefBase.ts:40]: @DanEastBentley @bsteinbk Is this OK?  Changing this forced me to update a lot of test snapshots...

**Comment by wgoehrig** [comment]
@grigasp @GerardasB we are now below UI and presentation minimum test coverage thresholds - likely due to changes in tsc's transpiled output.  Can we just lower these thresholds?  100% seems pretty silly anyways...

File: presentation/components/src/presentation-components/tree/controlled/WithFilteringSupport.tsx (hunk 6 above)
Comment (grigasp) [reply presentation/components/src/presentation-components/tree/controlled/WithFilteringSupport.tsx:87]: I think it's fine. @saskliutas what's your opinion?

**Review by grigasp — APPROVED** [review]

**Comment by grigasp** [comment]
> @grigasp @GerardasB we are now below UI and presentation minimum test coverage thresholds - likely due to changes in tsc's transpiled output. Can we just lower these thresholds?

Pushed a change that takes care of presentation coverage

File: presentation/components/src/presentation-components/tree/controlled/WithFilteringSupport.tsx (hunk 6 above)
Comment (saskliutas) [reply presentation/components/src/presentation-components/tree/controlled/WithFilteringSupport.tsx:87]: It should be fine.

**Comment by GerardasB** [comment]
> > @grigasp @GerardasB we are now below UI and presentation minimum test coverage thresholds - likely due to changes in tsc's transpiled output. Can we just lower these thresholds?
> 
> Pushed a change that takes care of presentation coverage

@wgoehrig UI coverage should also be good to go now.

File: core/geometry/src/test/geometry4d/MomentData.test.ts (hunk 4 above)
Comment (pmconne) [reply core/geometry/src/test/geometry4d/MomentData.test.ts:205]: `g` is always a `CurvePrimitive` here - compiler is correct.

**Review by pmconne — APPROVED** [review]

File: ui/framework/src/ui-framework/shared/ItemDefBase.ts (hunk 8 above)
Comment (bsteinbk) [reply ui/framework/src/ui-framework/shared/ItemDefBase.ts:40]: I am OK with the change as it seems harmless. It is sad that only Task and Workflow implement override getter/setter which triggered this change and I don't believe those classes are even used anywhere.

**Review by bsteinbk — APPROVED** [review]

File: core/geometry/src/test/geometry4d/MomentData.test.ts (hunk 4 above)
Comment (wgoehrig) [reply core/geometry/src/test/geometry4d/MomentData.test.ts:205]: Thanks Paul - I'll remove this.

**Review by calebmshafer — APPROVED** [review]

**Review by kabentley — APPROVED** [review]

**Review by scsewall — APPROVED** [review]

**Review by mgooding — APPROVED** [review]

**Review by swwilson-bsi — APPROVED** [review]

**Review by ramanujam-raman — APPROVED** [review]

File: core/geometry/src/test/serialization/BSIJSON.test.ts (hunk 5 above)
Comment (EarlinLutz) [reply core/geometry/src/test/serialization/BSIJSON.test.ts:69]: that's fine

**Review by EarlinLutz — APPROVED** [review]
In MomentData.test.ts, I added a call to pull in more test data so that the previously-unreachable CurveCollection block is hit.

File: core/geometry/src/test/geometry4d/MomentData.test.ts [hunk 9]
```diff
@@ -184,7 +184,7 @@ describe("MomentData", () => {
     expect(ck.getNumErrors()).equals(0);
   });
 
-  it("WireMoments", () => {
+  it.only("WireMoments", () => {
```
Comment (pmconne) [inline core/geometry/src/test/geometry4d/MomentData.test.ts:187]: Remove .only

**Review by pmconne — CHANGES_REQUESTED** [review]

File: core/geometry/src/test/geometry4d/MomentData.test.ts (hunk 9 above)
Comment (pmconne) [reply core/geometry/src/test/geometry4d/MomentData.test.ts:187]: @EarlinLutz

**Review by mdastous-bentley — APPROVED** [review]

File: core/geometry/src/test/geometry4d/MomentData.test.ts (hunk 9 above)
Comment (wgoehrig) [reply core/geometry/src/test/geometry4d/MomentData.test.ts]: Thanks @pmconne  - just pushed that change

**Review by RBBentley — APPROVED** [review]

**Review by pmconne — APPROVED** [review]
